okhttp/src/main/java/okhttp3/Headers.java

```diff
@@ -254,6 +253,46 @@ public static Headers of(Map<String, String> headers) {
     return new Headers(namesAndValues);
   }
 
+  /**
+   * Returns headers for the header names and values in the {@link Map}.
+   */
+  public static Headers fromMultimap(Map<String, List<String>> headers) {
+    if (headers == null) throw new NullPointerException("headers == null");
+
+    List<String> namesAndValues = new ArrayList<>(headers.size() * 2);
```

Comment:
would it be simpler to use the builder to reconstruct? You wouldn’t need to duplicate validation code

Reply:
sadly presizing the array list will be wrong if any header has multiple values

Reply:
>sadly presizing the array list will be wrong if any header has multiple values.
true, might be better to add some buffer or multiply by 3.

yeah, using the builder is nicer also because I can use only the public api.